zookeeper-server/src/test/java/org/apache/zookeeper/server/quorum/CommitProcessorMetricsTest.java

```diff
@@ -0,0 +1,483 @@
+        processRequestWithWait(createReadRequest(1L,3));
+
+        checkMetrics("requests_in_session_queue", 1L, 3L, 2D, 3L, 6);
+
+        commitWithWait(req1);
+
+        checkMetrics("requests_in_session_queue", 1L, 3L, 2.25D, 4L, 9);
+    }
+
+    @Test
+    public void testWriteFinalProcTime() throws Exception {
+        setupProcessors(0, 1000);
+
+        Request req1 = createWriteRequest(1l, 2);
+        processRequestWithWait(req1);
+
+        //no request sent to next processor yet
+        Map<String, Object> values = ServerMetrics.getAllValues();
+        Assert.assertEquals(0L, values.get("cnt_write_final_proc_time_ms"));
+
+        commitWithWait(req1);
+
+        values = ServerMetrics.getAllValues();
+        Assert.assertEquals(1L, values.get("cnt_write_final_proc_time_ms"));
+        checkTimeMetric((long)values.get("max_write_final_proc_time_ms"), 1000L, 2000L);
+    }
+
+    @Test
+    public void testReadFinalProcTime() throws Exception {
+        setupProcessors(0, 1000);
+
+        processRequestWithWait(createReadRequest(1L, 1));
+
+        Map<String, Object> values = ServerMetrics.getAllValues();
+        Assert.assertEquals(1L, values.get("cnt_read_final_proc_time_ms"));
+        checkTimeMetric((long)values.get("max_read_final_proc_time_ms"), 1000L, 2000L);
+    }
+
+    @Test
+    public void testCommitProcessTime() throws Exception {
+        setupProcessors(0, 0);
+        processRequestWithWait(createReadRequest(1L, 1));
+
+        Map<String, Object> values = ServerMetrics.getAllValues();
+        Assert.assertEquals(1L, values.get("cnt_commit_process_time"));
+        checkTimeMetric((long)values.get("max_commit_process_time"), 0L, 1000L);
+    }
+
+    @Test
+    public void testServerWriteCommittedTime() throws Exception {
+        setupProcessors(0, 0);
+        //a commit w/o pending request is a write from other servers
+        commitWithWait(createWriteRequest(1L, 1));
+
+        Map<String, Object> values = ServerMetrics.getAllValues();
+        Assert.assertEquals(1L, values.get("cnt_server_write_committed_time_ms"));
+        checkTimeMetric((long)values.get("max_server_write_committed_time_ms"), 0L, 1000L);
+    }
+
+    @Test
+    public void testLocalWriteCommittedTime() throws Exception {
+        setupProcessors(0, 0);
+        Request req1 = createWriteRequest(1l, 2);
+        processRequestWithWait(req1);
+        commitWithWait(req1);
+
+        Map<String, Object> values = ServerMetrics.getAllValues();
+
+        Assert.assertEquals(1L, values.get("cnt_local_write_committed_time_ms"));
+        checkTimeMetric((long)values.get("max_local_write_committed_time_ms"), 0l, 1000l);
+
+        Request req2 = createWriteRequest(1l, 2);
+        processRequestWithWait(req2);
+        //the second write will be stuck in the session queue for at least one second
+        //but the LOCAL_WRITE_COMMITTED_TIME is from when the commit is received
+        Thread.sleep(1000);
```

Comment:
This is going to lead to flaky test. If you cannot join threads properly here, I suggest to rather remove this validation.

Reply:
En... I don't think this particular sleep will lead to flaky test. The purpose of this sleep is not to wait for some other thread to finish something. It just delays the recording of metrics so I can check that the processing time is about 1 second instead of 0.